doc/dev/error-categorization.md

```diff
@@ -0,0 +1,173 @@
+
+##### Plugin Misbehavior ("C01330")
+
+`Plugin Misbehavior Errors` indicate that a plugin does not behave in an
+intended way. Unrecognized commands, unknown return codes, plugin creation
+errors, etc. belong to this category. Also uncaught exceptions belong here
+because Elektra expects all exceptions to be caught. For wrong plugin versions
+please use `Installation` errors. Users can try to remount, recompile the plugin
+under different options or use a different plugin (e.g., switching to `mini`
+instead of `ini`).
+
+### Conflicting State ("C02000")
+
+`Conflicting State Errors` are errors where the current state is incompatible
+with the attempted operation. These kind of errors are usually in resolver
+plugins when the state of the file has changed without the system knowing.
+Examples are the need for calling `kdbGet` before `kdbSet`. Another example
+would be to try to push your changes into a git repository where the remote
+branch has already changed. Try to synchronize your internal state and retry to
+get rid of this error.
+
+### Timeout ("C03000")
+
+`Timeout Errors` are errors which indicate temporary failures. This category
+should only be used if retries can safely be done without being harmful.
+Examples are lost connections to a server. Reactions are waiting for a short
+period and retry again.
+
+### Validation ("C04000")
+
+`Validation Errors` are heavily used for Elektra's `configuration specification`
+feature and should tell users that their given input does not match a certain
+pattern/type/expected semantic etc.
+
+Validation errors can either be syntactic or semantic.
+
+#### Syntactic ("C04100")
+
+`Syntactic Errors` are errors which tell users or applications that the current
+format is not valid. Examples are wrong date formats or missing closing brackets
+`]` inside of a regular expression when it is checked. Also path related errors
+like missing slashes come into this category. Parsing errors are also associated
+with syntactic errors such as unexpected encounters like missing line endings,
+illegal characters or wrong encodings. Also transformation and conversion errors
+are to be categorizes here because the format of the given input does not allow
+such actions. Users should try a different value and retry setting it with
+Elektra.
+
+#### Semantic ("C04200")
+
+`Semantic Errors` are errors which indicate a misunderstanding of the intended
+meaning between a user's/developer's/administrator's way of seeing a setting and
+the application's one. The main focus of this category is to enforce
+specifications compared to other categories. So if users provide input which do
+not match prespecified criterias even though syntactically everything is valid,
+this category should be used. Examples are references to non-existent keys
+(`reference` plugin), setting values to keys which require to be empty
+(`required` plugin), wrong provided values in a specification if you restricted
+the values (`enum` plugin), etc.
```

Comment:
Where are structural validation errors now?

Btw. you deviate here from literature. In literate semantic errors are only errors which represent domain-specific knowledge. Enum would be syntactically.

For users it might be important to be able to distinguish:

- errors where value(s) needs to be fixed to be valid again
- errors where key(s) need to be added/removed to be valid again

Reply:
> Btw. you deviate here from literature. In literate semantic errors are only errors which represent domain-specific knowledge. Enum would be syntactically.

I get your point! But are enum not also some sort of "domain-specific knowledge"? Eg. if you have a setting with `red`, `green`, `blue` as background setting and a user puts in `yellow`? The domain itself though is limited to these three colors.

Reply:
I do not see a clear separation from enum to regex or ranges. (As a matter of fact they can substitute each other in many cases.)